zipline-bytecode/src/main/kotlin/app/cash/zipline/bytecode/JsObjectWriter.kt

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.zipline.bytecode
+
+import okio.BufferedSink
+import okio.Closeable
+import okio.utf8Size
+
+/**
+ * Encodes a [JsObject] as bytes.
+ *
+ * @param atoms the mapping from string to integer used to encode the object. If the object was
+ *     decoded with a [JsObjectReader], it should be encoded with the same atoms.
+ */
+class JsObjectWriter(
+  private val atoms: AtomSet,
+  private val sink: BufferedSink
+) : Closeable by sink {
+  private var used: Boolean = false
+
+  fun writeJsObject(value: JsObject) {
+    check(!used)
+    used = true
```

Comment:
Would it make sense to just change `writeJsObject` and `readJsObject` into top-level functions instead of class methods? That way you don't need to defend against `JsObjectWriter` being used twice.

Reply:
Yeah I think that’s worth a shot. Let’s pair on that?

Reply:
Sure!